Enable Windows ProxyAll by default (#4980)
antrea-io/antrea: docs/windows.md

# Deploying Antrea on Windows

## Table of Contents

<!-- toc -->
- [Overview](#overview)
  - [Components that run on Windows](#components-that-run-on-windows)
  - [Antrea Windows demo](#antrea-windows-demo)
- [Deploying Antrea on Windows Worker Node](#deploying-antrea-on-windows-worker-node)
  - [Prerequisites](#prerequisites)

>   - [Installation as a Service (Containerd based runtimes)](#installation-as-a-service-containerd-based-runtimes)

Reviewer: when I read this section, it still mentioned running kube-proxy in userspace mode as a windows service (nssm). Should this section be updated as well?

Author: Updated.

  - [Installation as a Pod (Docker/containerd based runtimes)](#installation-as-a-pod-dockercontainerd-based-runtimes)
    - [Download & Configure Antrea for Linux](#download--configure-antrea-for-linux)
    - [Add Windows antrea-agent DaemonSet](#add-windows-antrea-agent-daemonset)
    - [Join Windows worker Nodes](#join-windows-worker-nodes)

>     - [1. (Optional) Install OVS (provided by Antrea or your own)](#1-optional-install-ovs-provided-by-antrea-or-your-own)

Author: Updated.

    - [2. Disable Windows Firewall](#2.-disable-windows-firewall)
    - [3. Install wins, kubelet, kubeadm and configure kubelet startup params](#3.-install-wins-kubelet-kubeadm-and-configure-kubelet-startup-params)
    - [4. Prepare Node environment needed by antrea-agent](#4.-prepare-node-environment-needed-by-antrea-agent)

>     - [5. Run kubeadm to join the Node](#5.-run-kubeadm-to-join-the-node)

Reviewer: hmmm these links don't seem to be working either?

Author: Updated all links, they are all available now.

    - [Verify your installation](#verify-your-installation)

>   - [Add Windows kube-proxy DaemonSet (only for Kubernetes versions prior to 1.26)](#add-windows-kube-proxy-daemonSet-only-for-kubernetes-versions-prior-to-1.26)

Reviewer: I don't know if it's a bug with a ToC generator or an issue on my side, but this link doesn't seem to be working.

### Installation as a Service (containerd based runtimes)

> First install Antrea (v0.13.0+ is required for containerd).The following

Reviewer 2: ```suggestion
First install Antrea (v0.13.0+ is required for containerd). The following
```

Author: @tnqn Updated, PTAL

### Installation as a Service (containerd based runtimes)

First install Antrea (v0.13.0+ is required for containerd).

> The following command deploys Antrea with the version specified by `<TAG>`:

Reviewer 3: Do not add a new line.

Author: Updated.

First install Antrea (v0.13.0+ is required for containerd).

> An example command that downloads and applies `antrea.yml` is like below:

Reviewer 3: I think no need to make a new paragraph for this sentence. Just merge it to the last paragraph.

Reviewer 3: You can say "The following command deploys Antrea v1.13.0" (or "... Antrea with the version specified by `<TAG>`") if you change `v1.13.0` to `<TAG>`).

Author: Updated.

```bash

> kubectl apply -f https://github.com/antrea-io/antrea/releases/download/v1.13.0/antrea.yml

Reviewer 3: I saw some other examples use <TAG> to express the release version. Could you use the same way in all examples?

Author: Sure, updated.

```bash
# Example:
kubectl apply -f https://github.com/antrea-io/antrea/releases/download/v1.13.0/antrea.yml
```

Then, you can run the following commands. [nssm](https://nssm.cc/) will install Antrea as a Windows service. Please ensure
`nssm` is on your machine, which is a handy tool to manage services on Windows. NOTE: `<KubernetesVersion>`, `<KubeconfigPath>`

> and `<KubeletKubeconfigPath>` should be set by you. `<KubeProxyKubeconfigPath>` is only for kube-proxy mode. E.g.

Author: Updated.

```powershell
$KubernetesVersion="v1.23.5"
$KubeConfig="C:/Users/Administrator/.kube/config" # admin kubeconfig
$KubeletKubeconfigPath="C:/etc/kubernetes/kubelet.conf"

> # $KubeProxyKubeconfigPath="C:/Users/Administrator/kubeproxy.conf" # Only for kube-proxy mode

Author: Added InstallKubeProxy to control the script, PTAL @tnqn

Then, you can run the following commands. [nssm](https://nssm.cc/) will install Antrea as a Windows service. Please ensure
`nssm` is on your machine, which is a handy tool to manage services on Windows. NOTE: `<KubernetesVersion>`, `<KubeconfigPath>`

> and `<KubeletKubeconfigPath>` should be set by you. `<KubeProxyKubeconfigPath>` is only for kube-proxy mode. E.g.

Reviewer: you should update the link above (`kubectl apply -f https://github.com/antrea-io/antrea/releases/download/v1.6.0/antrea.yml`). I think you may want to make it `v1.13.0` even though it's not released yet, as this is the first version that will support proxyAll for Windows?

Reviewer: and update `$TAG="v1.6.0"` below as well?

Author: Updated, thanks.

New-DirectoryIfNotExist "${AntreaHome}/logs"
New-DirectoryIfNotExist "${KubeProxyLogPath}"

> # Run kube-proxy

Author: Updated.

# nssm install kube-proxy "${KubernetesHome}/kube-proxy.exe" "--proxy-mode=userspace --kubeconfig=${KubeProxyKubeconfigPath} --log-dir=${KubeProxyLogPath} --logtostderr=false --alsologtostderr"
nssm install antrea-agent "${AntreaHome}/bin/antrea-agent.exe" "--config=${AntreaHome}/etc/antrea-agent.conf --logtostderr=false --log_dir=${AntreaHome}/logs --alsologtostderr --log_file_max_size=100 --log_file_max_num=4"

nssm set antrea-agent DependOnService ovs-vswitchd
# nssm set antrea-agent DependOnService kube-proxy ovs-vswitchd # Only for kube-proxy mode
nssm set antrea-agent Start SERVICE_DELAYED_AUTO_START

> # Start kube-proxy service

Author: Removed.

# Start-Service kube-proxy
Start-Service antrea-agent
```

### Installation as a Pod (Docker/containerd based runtimes)

> By default, Antrea will take over all the responsibilities of kube-proxy for Windows worker Nodes,

Author: Updated.

and kube-proxy should not be deployed on Windows worker Nodes to avoid conflicts with Antrea.

> For K8s versions prior to 1.26, Antrea can coexist on Windows Nodes with userspace kube-proxy.

Author: Updated.

New-DirectoryIfNotExist "${AntreaHome}/logs"
New-DirectoryIfNotExist "${KubeProxyLogPath}"
# Install kube-proxy service
# nssm install kube-proxy "${KubernetesHome}/kube-proxy.exe" "--proxy-mode=userspace --kubeconfig=${KubeProxyKubeconfigPath} --log-dir=${KubeProxyLogPath} --logtostderr=false --alsologtostderr"
nssm install antrea-agent "${AntreaHome}/bin/antrea-agent.exe" "--config=${AntreaHome}/etc/antrea-agent.conf --logtostderr=false --log_dir=${AntreaHome}/logs --alsologtostderr --log_file_max_size=100 --log_file_max_num=4"

nssm set antrea-agent DependOnService ovs-vswitchd
# nssm set antrea-agent DependOnService kube-proxy ovs-vswitchd # Only for kube-proxy mode
nssm set antrea-agent Start SERVICE_DELAYED_AUTO_START

# Start-Service kube-proxy
Start-Service antrea-agent
```

### Installation as a Pod (Docker/containerd based runtimes)

Starting with Antrea v1.13, Antrea will take over all the responsibilities of kube-proxy for Windows worker Nodes by default.

> Since kubernetes 1.26, kube-proxy should not be deployed on Windows worker Nodes due to deprecated userspace datapath.

Author: Updated.

> For K8s versions prior to 1.26, Antrea can coexist with userspace kube-proxy on Windows Nodes.

Author: Updated, thanks.

kubectl apply -f https://github.com/antrea-io/antrea/releases/download/<TAG>/antrea.yml
```

#### Add Windows kube-proxy DaemonSet (for Kubernetes versions prior to 1.26)

> Windows kube-proxy will no longer be supported by Antrea starting from Kubernetes 1.26. This decision is based on the deprecation of the userspace mode,

Reviewer 4: ```suggestion
Windows kube-proxy is no longer supported by Antrea since version 1.13. This decision is based on the deprecation of the kube-proxy userspace mode in Kubernetes,
```

Reviewer 4: But kuberenetes does not support kube-proxy any more, which kube-proxy userspace version would you suggest the users to use?

Author: After discussion, we still have a question regarding whether to maintain windows kube-proxy in document. While Antrea 1.13 sets "proxyall" as the default value for the Windows agent, it still maintains support for kube-proxy in Kubernetes versions prior to 1.26. Do you have any suggestions for updating the Windows documentation? @jianjuns @antoninbas

Reviewer: First, we should definitely enable the ToC for this document, as it is getting a bit confusing to navigate :)

Then, I think that we can still mention that mode (userspace kube-proxy) for the time being, but given that proxyAll also works for older K8s versions, it's probably better to have this information later in the document. Let's move it outside of the `### Installation as a Pod (Docker/Containerd based runtimes)` section.

This section (`### Installation as a Pod (Docker/Containerd based runtimes)`) should make the following clear:

> By default, Antrea will take over all the responsibilities of kube-proxy for Windows worker Nodes, and kube-proxy should not be deployed on Windows worker Nodes to avoid conflicts with Antrea. For K8s versions prior to 1.16, Antrea can coexist on Windows Nodes with user-space kube-proxy; for more information refer to section ...

You may want to put the new section right before `### Manually run kube-proxy and antrea-agent on Windows worker Nodes`.

> which Antrea heavily relies on. You should skip this step on the latest kubernetes cluster.

Reviewer 4: Add a sentence to explain that Antrea ProxyAll is used as a replacement of kube-proxy ?

Author: Updated.

New-DirectoryIfNotExist "${AntreaHome}/logs"
New-DirectoryIfNotExist "${KubeProxyLogPath}"
# Install kube-proxy service
if ($InstallKubeProxy) { nssm install kube-proxy "${KubernetesHome}/kube-proxy.exe" "--proxy-mode=userspace --kubeconfig=${KubeProxyKubeconfigPath} --log-dir=${KubeProxyLogPath} --logtostderr=false --alsologtostderr" }
nssm install antrea-agent "${AntreaHome}/bin/antrea-agent.exe" "--config=${AntreaHome}/etc/antrea-agent.conf --logtostderr=false --log_dir=${AntreaHome}/logs --alsologtostderr --log_file_max_size=100 --log_file_max_num=4"

nssm set antrea-agent DependOnService ovs-vswitchd
if ($InstallKubeProxy) { nssm set antrea-agent DependOnService kube-proxy ovs-vswitchd }
nssm set antrea-agent Start SERVICE_DELAYED_AUTO_START

if ($InstallKubeProxy) { Start-Service kube-proxy }
Start-Service antrea-agent
```

### Installation as a Pod (Docker/containerd based runtimes)

> Starting with Antrea v1.13, Antrea will take over all the responsibilities of kube-proxy for Windows nodes by default.

Reviewer 3: Please wrap lines at 80. Do not add a new line between sentences.

Author: Updated.

Since Kubernetes 1.26, kube-proxy should not be deployed on Windows Nodes with Antrea, as kube-proxy userspace mode is deprecated.
For Kubernetes versions prior to 1.26, Antrea can work with userspace kube-proxy on Windows Nodes.
For more information refer to section [Add Windows kube-proxy DaemonSet (only for Kubernetes versions prior to 1.26)](#add-windows-kube-proxy-daemonset-only-for-kubernetes-versions-prior-to-126)

Installing Antrea using [wins](https://github.com/rancher/wins) gives you a lot
of flexibility to manage it as a Pod if you are using Docker on Windows, this is
how you can run Antrea in a Pod.

#### Download & Configure Antrea for Linux

Deploy Antrea for Linux on the control-plane Node following [Getting started](getting-started.md)
document.

> The following command deploys Antrea with the version specified by `<TAG>`:

Reviewer 3: > I think no need to make a new paragraph for this sentence. Just merge it to the last paragraph.

Author: Updated.

New-DirectoryIfNotExist "${AntreaHome}/logs"
New-DirectoryIfNotExist "${KubeProxyLogPath}"
# Install kube-proxy service
if ($InstallKubeProxy) { nssm install kube-proxy "${KubernetesHome}/kube-proxy.exe" "--proxy-mode=userspace --kubeconfig=${KubeProxyKubeconfigPath} --log-dir=${KubeProxyLogPath} --logtostderr=false --alsologtostderr" }
nssm install antrea-agent "${AntreaHome}/bin/antrea-agent.exe" "--config=${AntreaHome}/etc/antrea-agent.conf --logtostderr=false --log_dir=${AntreaHome}/logs --alsologtostderr --log_file_max_size=100 --log_file_max_num=4"

nssm set antrea-agent DependOnService ovs-vswitchd
if ($InstallKubeProxy) { nssm set antrea-agent DependOnService kube-proxy ovs-vswitchd }
nssm set antrea-agent Start SERVICE_DELAYED_AUTO_START

if ($InstallKubeProxy) { Start-Service kube-proxy }
Start-Service antrea-agent
```

### Installation as a Pod (Docker/containerd based runtimes)

Starting with Antrea v1.13, Antrea will take over all the responsibilities of
kube-proxy for Windows nodes by default. Since Kubernetes 1.26, kube-proxy
should not be deployed on Windows Nodes with Antrea, as kube-proxy userspace
mode is deprecated.

> For Kubernetes versions prior to 1.26, Antrea can work with userspace kube-proxy

Reviewer 3: Do not add a new line. Wrap lines.

on Windows Nodes.
For more information refer to section [Add Windows kube-proxy DaemonSet (only for Kubernetes versions prior to 1.26)](#add-windows-kube-proxy-daemonset-only-for-kubernetes-versions-prior-to-126)

Installing Antrea using [wins](https://github.com/rancher/wins) gives you a lot
of flexibility to manage it as a Pod if you are using Docker on Windows, this is
how you can run Antrea in a Pod.

#### Download & Configure Antrea for Linux

Deploy Antrea for Linux on the control-plane Node following [Getting started](getting-started.md)
document. The following command deploys Antrea with the version specified by `<TAG>`:

```bash
kubectl apply -f https://github.com/antrea-io/antrea/releases/download/<TAG>/antrea.yml
```

#### Add Windows antrea-agent DaemonSet

> Starting from Antrea 1.13, you need to manually set the `kubeAPIServerOverride` field in the YAML configuration file as the Antrea Proxy `proxyAll` mode is enabled by default.

Reviewer 3: Wrap lines.

```yaml
    # Provide the address of Kubernetes apiserver, to override any value provided in kubeconfig or InClusterConfig.
    # Defaults to "". It must be a host string, a host:port pair, or a URL to the base of the apiserver.
    kubeAPIServerOverride: "10.10.1.1:6443"

    # Option antreaProxy contains AntreaProxy related configuration options.
    antreaProxy:
      # ProxyAll tells antrea-agent to proxy ClusterIP Service traffic, regardless of where they come from.
      # Therefore, running kube-proxy is no longer required. This requires the AntreaProxy feature to be enabled.
      # Note that this option is experimental. If kube-proxy is removed, option kubeAPIServerOverride must be used to access
      # apiserver directly.
      proxyAll: true
```

> An example command that downloads the antrea-agent manifest, sets kubeAPIServerOverride, and deploys the antrea-agent DaemonSet with Docker runtime is like below:

Reviewer 3: Wrap lines.

Author: Updated.

```bash
KUBE_APISERVER=$(kubectl config view -o jsonpath='{.clusters[0].cluster.server}') && \
curl -sL https://github.com/antrea-io/antrea/releases/download/<TAG>/antrea-windows.yml | \
sed "s|.*kubeAPIServerOverride: \"\"|    kubeAPIServerOverride: \"${KUBE_APISERVER}\"|g" | \
kubectl apply -f -
```

> Since Antrea 1.10, you can also deploy antrea-agent Windows DaemonSet with containerd runtime by applying file `antrea-windows-containerd.yml`.

Reviewer 3: Do not add a new line. Wrap lines.

Author: @jianjuns All updated, PTAL.

New-DirectoryIfNotExist "${AntreaHome}/logs"
New-DirectoryIfNotExist "${KubeProxyLogPath}"
# Install kube-proxy service
if ($InstallKubeProxy) { nssm install kube-proxy "${KubernetesHome}/kube-proxy.exe" "--proxy-mode=userspace --kubeconfig=${KubeProxyKubeconfigPath} --log-dir=${KubeProxyLogPath} --logtostderr=false --alsologtostderr" }
nssm install antrea-agent "${AntreaHome}/bin/antrea-agent.exe" "--config=${AntreaHome}/etc/antrea-agent.conf --logtostderr=false --log_dir=${AntreaHome}/logs --alsologtostderr --log_file_max_size=100 --log_file_max_num=4"

nssm set antrea-agent DependOnService ovs-vswitchd
if ($InstallKubeProxy) { nssm set antrea-agent DependOnService kube-proxy ovs-vswitchd }
nssm set antrea-agent Start SERVICE_DELAYED_AUTO_START

> # Start-Service kube-proxy

Author: Added InstallKubeProxy.

Start-Service antrea-agent
```

### Installation as a Pod (Docker/containerd based runtimes)

Starting with Antrea v1.13, Antrea will take over all the responsibilities of kube-proxy for Windows nodes by default.
Since Kubernetes 1.26, kube-proxy should not be deployed on Windows Nodes with Antrea, as kube-proxy userspace mode is deprecated.
For Kubernetes versions prior to 1.26, Antrea can work with userspace kube-proxy on Windows Nodes.

> For more information refer to section [Add Windows kube-proxy DaemonSet (only for Kubernetes versions prior to 1.26)](#add-windows-kube-proxy-daemonSet-only-for-kubernetes-versions-prior-to-1.26)

Reviewer 2: the anchor is invalid

### Installation as a Pod (Docker/containerd based runtimes)

By default, Antrea will take over all the responsibilities of kube-proxy for Windows worker Nodes,
and kube-proxy should not be deployed on Windows worker Nodes to avoid conflicts with Antrea.

> For K8s versions prior to 1.26, Antrea can coexist on Windows Nodes with user-space kube-proxy.

Reviewer: s/user-space/userspace

> Configure the Antrea for Linux on the control-plane Node following [Getting started](getting-started.md)

Reviewer 3: "Configure the Antrea" -> "Deploy Antrea"

Author: Updated.

document.

> An example command that downloads and applies `antrea.yml` is like below:

Reviewer 3: Ditto

Reviewer 3: `antrea.yml` -> the Antrea deployment manifest

    kubeAPIServerOverride: "10.10.1.1:6443"

    # Option antreaProxy contains AntreaProxy related configuration options.
    antreaProxy:
      # ProxyAll tells antrea-agent to proxy ClusterIP Service traffic, regardless of where they come from.
      # Therefore, running kube-proxy is no longer required. This requires the AntreaProxy feature to be enabled.
      # Note that this option is experimental. If kube-proxy is removed, option kubeAPIServerOverride must be used to access
      # apiserver directly.
      proxyAll: true
```

An example command that downloads the antrea-agent manifest, sets kubeAPIServerOverride, and deploys the antrea-agent DaemonSet with Docker runtime is like below:

```bash
KUBE_APISERVER=$(kubectl config view -o jsonpath='{.clusters[0].cluster.server}') && \
curl -sL https://github.com/antrea-io/antrea/releases/download/<TAG>/antrea-windows.yml | \
sed "s|.*kubeAPIServerOverride: \"\"|    kubeAPIServerOverride: \"${KUBE_APISERVER}\"|g" | \
kubectl apply -f -
```

Since Antrea 1.10, you can also deploy antrea-agent Windows DaemonSet with containerd runtime by applying file `antrea-windows-containerd.yml`.

> An example command that downloads the antrea-agent manifest, sets kubeAPIServerOverride, and deploys the antrea-agent DaemonSet with containerd runtime is like below:

Reviewer 3: Ditto

kubectl apply -f https://github.com/antrea-io/antrea/releases/download/<TAG>/antrea.yml
```

#### Add Windows antrea-agent DaemonSet

> Starting from Antrea 1.13, you need to manually set the `kubeAPIServerOverride` field in the YAML configuration file as the proxy-all mode is enabled by default.

Reviewer 3: Could we automate this step if we use Powershell for Windows installation anyway? @wenyingd 

Do not mean we need to do in this PR.

Reviewer 4: We have added the command to modify the configuration in powershell script "Start-AntreaAgent.ps1" now, but the script is only working when running antrea-agent as Windows service. We now don't have automation to update in antrea-agent-windows-containerd.yml which is applied to run antrea-agent Pod on Windows. 

If we plan to add support for such automation, shall we assume users to run the script to on Windows or Linux? The CI testbed now applies the yml on master (Linux) Node.

Reviewer 3: Ok, if we do not have script for containerd, we can keep it manual. Could we make a single command (with pipe) to download the yaml and change the parameter and even apply the yaml?

Author: I added a single command to apply yaml, does it look good to you? `curl -sL https://github.com/antrea-io/antrea/releases/download/<TAG>/antrea-windows-containerd.yml | sed "s|.*kubeAPIServerOverride: \"\"|    kubeAPIServerOverride: \"$(kubectl --kubeconfig=$KubeConfigFile config view -o jsonpath='{.clusters[0].cluster.server}')\"|g" | kubectl apply -f -`

> Starting from Antrea 1.13, you need to manually set the `kubeAPIServerOverride` field in the YAML configuration file as the proxy-all mode is enabled by default.

Author: Updated.

kubectl apply -f https://github.com/antrea-io/antrea/releases/download/<TAG>/antrea.yml
```

#### Add Windows antrea-agent DaemonSet

Starting from Antrea 1.13, you need to manually set the `kubeAPIServerOverride` field in the YAML configuration file as the Antrea Proxy `proxyAll` mode is enabled by default.

```yaml
# Provide the address of Kubernetes apiserver, to override any value provided in kubeconfig or InClusterConfig.
    # Defaults to "". It must be a host string, a host:port pair, or a URL to the base of the apiserver.
    kubeAPIServerOverride: "10.10.1.1:6443"

    # Option antreaProxy contains AntreaProxy related configuration options.
    antreaProxy:
      # ProxyAll tells antrea-agent to proxy ClusterIP Service traffic, regardless of where they come from.
      # Therefore, running kube-proxy is no longer required. This requires the AntreaProxy feature to be enabled.
      # Note that this option is experimental. If kube-proxy is removed, option kubeAPIServerOverride must be used to access
      # apiserver directly.
      proxyAll: true
```

```bash
# Example:

> curl -sL https://github.com/antrea-io/antrea/releases/download/<TAG>/antrea-windows-containerd.yml | sed "s|.*kubeAPIServerOverride: \"\"|    kubeAPIServerOverride: \"$(kubectl --kubeconfig=$KubeConfigFile config view -o jsonpath='{.clusters[0].cluster.server}')\"|g" | kubectl apply -f -

Reviewer 3: What does "$KubeConfigFile" points to? Do we need users to provide it?

I feel better to make it a separate command to get the apiserver endpoint with:
kubectl --kubeconfig=$KubeConfigFile config view -o jsonpath='{.clusters[0].cluster.server}'. Maybe set the value to a variable.

Author: Yes we can remove KubeConfigFile here. I added a separate command to get apiserver endpoint.

kubectl apply -f https://github.com/antrea-io/antrea/releases/download/<TAG>/antrea.yml
```

#### Add Windows antrea-agent DaemonSet

Starting from Antrea 1.13, you need to manually set the `kubeAPIServerOverride` field in the YAML configuration file as the Antrea Proxy `proxyAll` mode is enabled by default.

```yaml

> # Provide the address of Kubernetes apiserver, to override any value provided in kubeconfig or InClusterConfig.

Author: Updated.

    # Defaults to "". It must be a host string, a host:port pair, or a URL to the base of the apiserver.
    kubeAPIServerOverride: "10.10.1.1:6443"

    # Option antreaProxy contains AntreaProxy related configuration options.
    antreaProxy:
      # ProxyAll tells antrea-agent to proxy ClusterIP Service traffic, regardless of where they come from.
      # Therefore, running kube-proxy is no longer required. This requires the AntreaProxy feature to be enabled.
      # Note that this option is experimental. If kube-proxy is removed, option kubeAPIServerOverride must be used to access
      # apiserver directly.
      proxyAll: true
```

```bash
# Example:

> KUBE_API_SERVER=$(kubectl config view -o jsonpath='{.clusters[0].cluster.server}')

Reviewer 3: How about add "&&" to the end to connect the curl command?

> KUBE_API_SERVER=$(kubectl config view -o jsonpath='{.clusters[0].cluster.server}')

Author: Updated.

> curl -sL https://github.com/antrea-io/antrea/releases/download/<TAG>/antrea-windows-containerd.yml | sed "s|.*kubeAPIServerOverride: \"\"|    kubeAPIServerOverride: \"${KUBE_API_SERVER}\"|g" | kubectl apply -f -

Reviewer 3: The line is too long. Use "\" to wrap the lines after each "|".

Author: Sure, updated.

kubectl apply -f https://github.com/antrea-io/antrea/releases/download/<TAG>/antrea.yml
```

#### Add Windows antrea-agent DaemonSet

Starting from Antrea 1.13, you need to manually set the `kubeAPIServerOverride` field in the YAML configuration file as the Antrea Proxy `proxyAll` mode is enabled by default.

```yaml
    # Provide the address of Kubernetes apiserver, to override any value provided in kubeconfig or InClusterConfig.
    # Defaults to "". It must be a host string, a host:port pair, or a URL to the base of the apiserver.
    kubeAPIServerOverride: "10.10.1.1:6443"

    # Option antreaProxy contains AntreaProxy related configuration options.
    antreaProxy:
      # ProxyAll tells antrea-agent to proxy ClusterIP Service traffic, regardless of where they come from.
      # Therefore, running kube-proxy is no longer required. This requires the AntreaProxy feature to be enabled.
      # Note that this option is experimental. If kube-proxy is removed, option kubeAPIServerOverride must be used to access
      # apiserver directly.
      proxyAll: true
```

```bash

> # Example:

Reviewer 3: I feel it would be better to remove the "Example:" comment from the command, but add a sentence before the command, like: "An example command that downloads the antrea-agent manifest, sets `kubeAPIServerOverride`, and deploys the antrea-agent DaemonSet is like below:".

So, readers can directly copy the command by clicking the copy button on the web page.

Could you also change other commands with the "Example:" comment?

Author: Sure, removed all `example` comment and updated sentences.

KUBE_APISERVER=$(kubectl config view -o jsonpath='{.clusters[0].cluster.server}') && \
curl -sL https://github.com/antrea-io/antrea/releases/download/<TAG>/antrea-windows-containerd.yml | \
sed "s|.*kubeAPIServerOverride: \"\"|    kubeAPIServerOverride: \"${KUBE_APISERVER}\"|g" | \
kubectl apply -f -
```

> Now you can deploy antrea-agent Windows DaemonSet with Docker runtime by applying file `antrea-windows.yml`.

Reviewer 3: Do not understand the following contents. The previous command already deploys antrea-agent right? Why the following are still talking about antrea-agent deployment. Could you check?

Author: Updated, since 1.13 I think we should remove commands without updating kubeAPIServerOverride because the default proxyall value is true.

kubectl apply -f https://github.com/antrea-io/antrea/releases/download/<TAG>/antrea-windows.yml
```

> Since Antrea 1.10, you can also deploy antrea-agent Windows DaemonSet with Containerd runtime by

Author: Udpated.

get-service ovs-vswitchd
```

##### 2. Disable Windows Firewall

```powershell
Set-NetFirewallProfile -Profile Domain,Public,Private -Enabled False
```

##### 3. Install wins, kubelet, kubeadm and configure kubelet startup params

Firstly, install wins, kubelet, kubeadm using script `PrepareNode.ps1` provided
by kubernetes. The third component [`wins`](https://github.com/rancher/wins) is
used to run kube-proxy and antrea-agent on Windows host inside the Windows
container.

> An example command that downloads and executes `PrepareNode.ps1` is like below:

Reviewer 3: Ditto

#### Add Windows antrea-agent DaemonSet

> Starting from Antrea 1.13, you need to manually set the kubeAPIServerOverride field in the YAML configuration file as the proxy-all mode is enabled by default.

Reviewer 4: ```suggestion
Starting from Antrea 1.13, you need to manually set the `kubeAPIServerOverride` field in the YAML configuration file as the proxy-all mode is enabled by default.
```

> 

Reviewer 4: Give an example of antrea-agent-config snippet to show how to update the field?

Author: Do you mean give command to get the apiserver address here?

Reviewer 4: No, I mean you should gve an example to show that "kubeAPIServerride" is configured.

Author: Updated.

```powershell
mkdir c:\k\antrea
cd c:\k\antrea
$TAG="v1.13.0"
curl.exe -LO https://raw.githubusercontent.com/antrea-io/antrea/${TAG}/hack/windows/Clean-AntreaNetwork.ps1
curl.exe -LO https://raw.githubusercontent.com/antrea-io/antrea/${TAG}/hack/windows/Prepare-ServiceInterface.ps1
curl.exe -LO https://raw.githubusercontent.com/antrea-io/antrea/${TAG}/hack/windows/Prepare-AntreaAgent.ps1
.\Prepare-AntreaAgent.ps1
```

The script `Prepare-AntreaAgent.ps1` performs following tasks:

> * Prepare network adapter for kube-proxy.(for Kubernetes versions prior to 1.26)

Author: Updated.

    kube-proxy needs a network adapter to configure Kubernetes Services IPs and
    uses the adapter for proxying connections to Service. Use following script
    to create the network adapter. The adapter will be deleted automatically by
    Windows after the Windows Node reboots.

    Starting from Kubernetes 1.26, the kube-proxy userspace datapath is no longer

>     supported. For clusters running a version higher than 1.26, you should exclude

Author: Yes, it's better to use skip here, updated.

The script `Prepare-AntreaAgent.ps1` performs following tasks:

* Prepare network adapter for kube-proxy.(for Kubernetes versions prior to 1.26)

    kube-proxy needs a network adapter to configure Kubernetes Services IPs and
    uses the adapter for proxying connections to Service. Use following script
    to create the network adapter. The adapter will be deleted automatically by
    Windows after the Windows Node reboots.

    Starting from Kubernetes 1.26, the kube-proxy userspace datapath is no longer
    supported. For clusters running a version higher than 1.26, you should exclude
    the preparation of the kube-proxy network adapter by executing the command

>     .\Prepare-AntreaAgent.ps1 -InstallKubeProxy $false.

Reviewer: why aren't we making this the default behavior?

Author: I'm concerned if users refer to the documentation from a previous release, the script's default behavior may be different. If you don't see this as a problem, I can go ahead and change its default value. @antoninbas

Reviewer: I understand what you mean. I think the problem may stem from the fact that the links to download the scripts are using the `main` branch:
```pwsh
curl.exe -LO https://raw.githubusercontent.com/antrea-io/antrea/main/hack/windows/Clean-AntreaNetwork.ps1
curl.exe -LO https://raw.githubusercontent.com/antrea-io/antrea/main/hack/windows/Prepare-ServiceInterface.ps1
curl.exe -LO https://raw.githubusercontent.com/antrea-io/antrea/main/hack/windows/Prepare-AntreaAgent.ps1
```
I think we can keep the `-InstallKubeProxy` as you currently have it. However, we should do something about these links so we don't have the same problem in the future.
We could do the same thing as for the `Helper.psm1` link earlier in the document:
```pwsh
curl.exe -LO "https://raw.githubusercontent.com/antrea-io/antrea/${TAG}/hack/windows/Helper.psm1"
```
We could also package these scripts as one zip file, and add it as a release asset. Not sure what the best solution is. The first one is probably the easiest, and we could make that change as part of this PR.

Author: Yes, we can update related links first.

>     .\Prepare-AntreaAgent.ps1 -InstallKubeProxy $false.

Reviewer: command should be quoted with backticks

Register-ScheduledJob -Name PrepareAntreaAgent -Trigger $trigger  -ScriptBlock { Invoke-Expression C:\k\antrea\Prepare-AntreaAgent.ps1 } -ScheduledJobOption $options
```

##### 5. Run kubeadm to join the Node

On Windows Node, run the `kubeadm join` command to join the cluster. The token
is provided by the control-plane Node.

> 

Reviewer 3: I feel better to merge the three paragraphs (line 313 to 320) to one.

Author: Merged into one.

`kubeadm token create --print-join-command` (on the control-plane Node) to
generate a new token and join command.

> An example of kubeadm join command is like below:

Reviewer 3: Ditto

Author: Updated.

> An example of kubeadm join command is like below:

Reviewer 3: kubeadmin join -> `kubeadmin join`

Reviewer 3: Remove "of".

kube-proxy-windows-2d45w                               1/1     Running     0          102s
```

### Add Windows kube-proxy DaemonSet (only for Kubernetes versions prior to 1.26)

Windows kube-proxy will no longer be supported by Antrea starting from Kubernetes 1.26. This decision is based on the deprecation of the userspace mode,
which Antrea heavily relies on. You should skip this step on the latest kubernetes cluster and enable proxy-all mode for windows agent.

Add Windows-compatible versions of kube-proxy by applying file `kube-proxy.yaml`.

Download `kube-proxy.yaml` from kubernetes official repository and set
kube-proxy version.

```bash
# Example:

> curl -L "https://github.com/kubernetes-sigs/sig-windows-tools/releases/download/v0.1.5/kube-proxy.yml" | sed 's/VERSION/v1.23.5/g' > kube-proxy.yml

Reviewer: maybe use the latest supported version instead of `v1.23.5`?

Author: I found the yaml file has been updated and the image is VERSION-nanoserver now, which doesn't support our current kube-proxy script, the latest available image version is v1.20.0.

kube-proxy-windows-2d45w                               1/1     Running     0          102s
```

### Add Windows kube-proxy DaemonSet (only for Kubernetes versions prior to 1.26)

> Windows kube-proxy will no longer be supported by Antrea starting from Kubernetes 1.26. This decision is based on the deprecation of the userspace mode,

Author: Updated.

> Starting from Kubernetes 1.26, Antrea will no longer support Windows kube-proxy due to its lack of support for the kernel datapath, which Antrea requires.

Author: Updated.

> You should skip this step on the kubernetes cluster starting from 1.26 and enable proxy-all mode for windows agent.

Author: Updated

> which Antrea heavily relies on. You should skip this step on the latest kubernetes cluster and enable proxy-all mode for windows agent.

Author: Updated.

> which Antrea heavily relies on. You should skip this step on the latest kubernetes cluster and enable proxy-all mode for windows agent.

Author: Added 1.26 version info.

> Add Windows-compatible versions of kube-proxy by applying file `kube-proxy.yaml`.

Reviewer 3: Remove "file" or change to "manifest"

> Download `kube-proxy.yaml` from kubernetes official repository and set

Reviewer 3: the Kubernetes

> Download `kube-proxy.yaml` from kubernetes official repository and set

Author: Updated.

kube-proxy version.

```bash
# Example:
curl -L "https://github.com/kubernetes-sigs/sig-windows-tools/releases/download/v0.1.5/kube-proxy.yml" | sed 's/VERSION-nanoserver/v1.20.0/g' > kube-proxy.yml
```

> Replace the content of `run-script.ps1` in configmap named `kube-proxy-windows`

Reviewer 3: In the "kube-proxy-windows" ConfigMap

> Replace the content of `run-script.ps1` in configmap named `kube-proxy-windows`

Reviewer 3: Can we automate this part with a Powershell script, and even auto-download kube-proxy.yml? It looks complicated.

@wenyingd

Reviewer 4: Since we don't plan to support kube-proxy userspace mode any more, maybe it is not worth to add automation for changes with kube-proxy on Windows?

Reviewer 3: Sure, we can keep it manual then.

kind: ConfigMap
metadata:
  labels:
    app: kube-proxy
  name: kube-proxy-windows
  namespace: kube-system
```

For containerd runtime, replace the content of `run-script.ps1` with following:

```yaml
apiVersion: v1
data:
  run-script.ps1: |-
    $mountPath = $env:CONTAINER_SANDBOX_MOUNT_POINT
    $mountPath =  ($mountPath.Replace('\', '/')).TrimEnd('/') 
    New-Item -Path "c:/var/lib" -Name "kube-proxy" -ItemType "directory" -Force
    ((Get-Content -path $mountPath/var/lib/kube-proxy/kubeconfig.conf -Raw) -replace '/var',"$($mountPath)/var") | Set-Content -Path /var/lib/kube-proxy/kubeconfig.conf
    ((Get-Content -path /var/lib/kube-proxy/kubeconfig.conf -Raw) -replace '\/',"/") | Set-Content -Path /var/lib/kube-proxy/kubeconfig.conf
    sed -i 's/mode: iptables/mode: \"\"/g' $mountPath/var/lib/kube-proxy/config.conf
    & "$mountPath/k/kube-proxy/kube-proxy.exe" --config=$mountPath/var/lib/kube-proxy/config.conf --v=10 --proxy-mode=userspace --hostname-override=$env:NODE_NAME
kind: ConfigMap
metadata:
  labels:
    app: kube-proxy
  name: kube-proxy-windows
  namespace: kube-system
```

> Set the `hostNetwork` option as true in spec of kube-proxy-windows daemonset.

Reviewer 3: as true -> to `true`

Author: Updated.

```yaml
apiVersion: apps/v1
kind: DaemonSet
metadata:
  labels:
    k8s-app: kube-proxy
  name: kube-proxy-windows
  namespace: kube-system
spec:
  selector:
    matchLabels:
      k8s-app: kube-proxy-windows
  template:
    metadata:
      labels:
        k8s-app: kube-proxy-windows
    spec:
      hostNetwork: true
```

> For containerd runtime, Set `hostNetwork` as true and add options in spec and command.

Reviewer 3: as true -> to `true`

> For containerd runtime, Set `hostNetwork` to true and add options in spec and command.

Reviewer 3: true -> `true`

Author: Updated.

kube-proxy-windows-2d45w                               1/1     Running     0          102s
```

### Add Windows kube-proxy DaemonSet (only for Kubernetes versions prior to 1.26)

Starting from Kubernetes 1.26, Antrea will no longer support Windows kube-proxy
because kube-proxy userspace mode is deprecated and kernelspace mode does not
work with Antrea. For a cluster with Kubernetes version 1.26 or later, please
refer to [Deploying Antrea on Windows Worker Node](#Deploying-Antrea-on-Windows-Worker-Node)

First, deploy a Windows-compatible version of kube-proxy. You can download `kube-proxy.yaml`
from the Kubernetes github repository to deploy kube-proxy. The kube-proxy version
in the YAML file must be set to a Windows compatible version.

> The following command downloads and applies `kube-proxy.yml`:

Reviewer 3: Do not add a new line.

Author: Updated.

kube-proxy-windows-2d45w                               1/1     Running     0          102s
```

### Add Windows kube-proxy DaemonSet (only for Kubernetes versions prior to 1.26)

Starting from Kubernetes 1.26, Antrea will no longer support Windows kube-proxy
because kube-proxy userspace mode is deprecated and kernelspace mode does not

> work with Antrea. You should skip this step on the kubernetes cluster starting

Reviewer 3: Should we say: "For a cluster with Kubernetes version 1.26 or later, please refer to [Deploying Antrea on Windows Worker Node]"?

Author: Updated.

from 1.26 and enable Antrea proxy `proxyAll` mode for Windows agent.

> Add Windows-compatible versions of kube-proxy by applying `kube-proxy.yaml`.

Reviewer 3: "First, deploy a Windows-compatible version of kube-proxy."?

Download `kube-proxy.yaml` from the Kubernetes official repository and set

> the kube-proxy version.

Reviewer 3: "You can download `kube-proxy.yaml` from the Kubernetes github repository to deploy kube-proxy. The kube-proxy version in the YAML file must be set to a Windows compatible version"?

Author: Updated.

> An example command that downloads and applies `kube-proxy.yml` is like below:

Reviewer 3: Ditto

Reviewer 3: The following command deploys kube-proxy v1.20.0?

Author: Yes, it's the latest compatible kubeproxy image we can use.

kind: ConfigMap
metadata:
  labels:
    app: kube-proxy
  name: kube-proxy-windows
  namespace: kube-system
```

For containerd runtime, replace the content of `run-script.ps1` with following:

```yaml
apiVersion: v1
data:
  run-script.ps1: |-
    $mountPath = $env:CONTAINER_SANDBOX_MOUNT_POINT
    $mountPath =  ($mountPath.Replace('\', '/')).TrimEnd('/') 
    New-Item -Path "c:/var/lib" -Name "kube-proxy" -ItemType "directory" -Force
    ((Get-Content -path $mountPath/var/lib/kube-proxy/kubeconfig.conf -Raw) -replace '/var',"$($mountPath)/var") | Set-Content -Path /var/lib/kube-proxy/kubeconfig.conf
    ((Get-Content -path /var/lib/kube-proxy/kubeconfig.conf -Raw) -replace '\/',"/") | Set-Content -Path /var/lib/kube-proxy/kubeconfig.conf
    sed -i 's/mode: iptables/mode: \"\"/g' $mountPath/var/lib/kube-proxy/config.conf
    & "$mountPath/k/kube-proxy/kube-proxy.exe" --config=$mountPath/var/lib/kube-proxy/config.conf --v=10 --proxy-mode=userspace --hostname-override=$env:NODE_NAME
kind: ConfigMap
metadata:
  labels:
    app: kube-proxy
  name: kube-proxy-windows
  namespace: kube-system
```

> Set the `hostNetwork` option to `true` in spec of kube-proxy-windows daemonset.

Reviewer 3: DaemonSet

Probably "in the kube-proxy-windows DaemonSet spec"

```yaml
apiVersion: apps/v1
kind: DaemonSet
metadata:
  labels:
    k8s-app: kube-proxy
  name: kube-proxy-windows
  namespace: kube-system
spec:
  selector:
    matchLabels:
      k8s-app: kube-proxy-windows
  template:
    metadata:
      labels:
        k8s-app: kube-proxy-windows
    spec:
      hostNetwork: true
```

> For containerd runtime, Set `hostNetwork` to `true` and add options in spec and command.

Reviewer 3: "For containerd runtime, also add the following to the kube-proxy-windows DaemonSet spec"?

Author: Updated.

mkdir c:\k\antrea
cd c:\k\antrea
curl.exe -LO https://github.com/antrea-io/antrea/releases/download/<TAG>/Start-AntreaAgent.ps1
# Run antrea-agent without kube-proxy
# $KubeConfigPath is the path of kubeconfig file

> ./Start-AntreaAgent.ps1 -kubeconfig $KubeConfigPath -StartKubeProxy $false

Reviewer 4: Why shall we run twice with "Start-AntreaAgent.ps1"  ( line 454 and line 456 )

Author: To indicate the different command which enables kube-proxy.

Reviewer 4: If it is meaning that two command examples with different option values, maybe move the second one (requiring kube-proxy) into commented lines?

Author: Updated.

# Run Antrea-Agent with kube-proxy (deprecated since Kubernetes 1.26)

> ./Start-AntreaAgent.ps1 -kubeconfig $KubeConfigPath

Reviewer 4: After your patch, the default value of `$StartKubeProxy` is "$false", so line 456 actually does not start kube-proxy.

Author: Updated.
